File: apps/helper/src/rename.ts
const VIDEO_EXT = new Set([
  '.mkv',
  '.mp4',
  '.avi',
  '.ts',
  '.m2ts',
  '.mts',
  '.webm',
  '.mov',
  '.wmv',
  '.flv',
])

const SUB_EXT = new Set([
  '.srt',
  '.ass',
  '.ssa',
  '.sub',
  '.vtt',
  '.idx',
  '.sup',
])

const EXTRA_TOKEN = /(?:^|[\\/._\-\s])(?:sample|ncop|nced)(?:[\\/._\-\s]|$)/i

export function formatEpisodeName(
  title: string,
  season: number,
  episode: number,
): string {
  return `${title} - S${pad2(season)}E${pad2(episode)}`
}

export function formatSavePath(
  libraryRoot: string,
  title: string,
  season: number,
): string {
  const seasonDir = `Season ${pad2(season)}`
  const safeTitle = sanitizePathSegment(title)
  if (!libraryRoot) {
    return `${safeTitle}/${seasonDir}`
  }
  const sep
    = libraryRoot.includes('\\') && !libraryRoot.includes('/') ? '\\' : '/'
  return `${trimSlash(libraryRoot)}${sep}${safeTitle}${sep}${seasonDir}`
}

export function isSkippedExtra(filePath: string): boolean {
  return filePath.split(/[/\\]/).some(part => EXTRA_TOKEN.test(part))
}

export function planEpisodeRenames(input: {
  displayName: string
  files: Array<{ name: string, size: number }>
}): Array<{ from: string, to: string }> {
  const usable = input.files.filter(file => !isSkippedExtra(file.name))
  const videos = usable.filter(file => VIDEO_EXT.has(extname(file.name)))
  if (videos.length === 0) {
    return []
  }

  const primary = videos.reduce((best, file) =>
    file.size > best.size ? file : best)
  const primaryStem = stem(primary.name)
  const dir = dirname(primary.name)
  const plans: Array<{ from: string, to: string }> = []

  const videoTo = joinDir(dir, `${input.displayName}${extname(primary.name)}`)
  if (primary.name !== videoTo) {
    plans.push({ from: primary.name, to: videoTo })
  }

  for (const file of usable) {
    if (!SUB_EXT.has(extname(file.name))) {
      continue
    }
    const subStem = stem(file.name)
    if (subStem !== primaryStem && !subStem.startsWith(primaryStem)) {
      continue
    }
    const extra = subStem.slice(primaryStem.length)
    const subTo = joinDir(
      dir,
      `${input.displayName}${extra}${extname(file.name)}`,
    )
    if (file.name !== subTo) {
      plans.push({ from: file.name, to: subTo })
    }
  }

  return plans
}

function pad2(value: number): string {
  return String(value).padStart(2, '0')
}

function sanitizePathSegment(name: string): string {
  return name.replaceAll(/[\\/:*?"<>|]/g, '_').trim()
}

function trimSlash(path: string): string {
  return path.replace(/[/\\]+$/, '')
}

function extname(filePath: string): string {
  const base = basename(filePath)
  const index = base.lastIndexOf('.')
  return index >= 0 ? base.slice(index).toLowerCase() : ''
}

function basename(filePath: string): string {
  const index = Math.max(filePath.lastIndexOf('/'), filePath.lastIndexOf('\\'))
  return index >= 0 ? filePath.slice(index + 1) : filePath
}

function dirname(filePath: string): string {
  const index = Math.max(filePath.lastIndexOf('/'), filePath.lastIndexOf('\\'))
  return index >= 0 ? filePath.slice(0, index) : ''
}

function stem(filePath: string): string {
  const base = basename(filePath)
  const index = base.lastIndexOf('.')
  return index >= 0 ? base.slice(0, index) : base
}

function joinDir(dir: string, name: string): string {
  if (!dir) {
    return name
  }
  const sep = dir.includes('\\') && !dir.includes('/') ? '\\' : '/'
  return `${dir}${sep}${name}`
}

File: apps/helper-go/internal/rename/rename.go
package rename

import (
	"fmt"
	"path"
	"regexp"
	"strings"

	"github.com/Torrent-Vibe/Torrent-Vibe/apps/helper-go/internal/qb"
)

type Plan struct {
	From string
	To   string
}

var (
	videoExt = map[string]struct{}{
		".mkv": {}, ".mp4": {}, ".avi": {}, ".ts": {}, ".m2ts": {},
		".mts": {}, ".webm": {}, ".mov": {}, ".wmv": {}, ".flv": {},
	}
	subExt = map[string]struct{}{
		".srt": {}, ".ass": {}, ".ssa": {}, ".sub": {}, ".vtt": {}, ".idx": {}, ".sup": {},
	}
	extraToken = regexp.MustCompile(`(?i)(^|[\\/._\-\s])(sample|ncop|nced|extras?)($|[\\/._\-\s])`)
	unsafeSeg  = regexp.MustCompile(`[\\/:*?"<>|]`)
)

func IsVideo(filePath string) bool {
	_, ok := videoExt[extname(filePath)]
	return ok
}

func IsSkippedExtra(filePath string) bool {
	return isSkippedExtra(filePath)
}

func SanitizeTitle(title string) string {
	safe := unsafeSeg.ReplaceAllString(title, "_")
	safe = strings.TrimSpace(safe)
	return strings.TrimRight(safe, " .")
}

func FormatEpisodeName(title string, season, episode int) string {
	return fmt.Sprintf("%s - S%02dE%02d", title, season, episode)
}

func FormatSavePath(libraryRoot, title string, season int) string {
	seasonDir := fmt.Sprintf("Season %02d", season)
	safeTitle := unsafeSeg.ReplaceAllString(title, "_")
	safeTitle = strings.TrimSpace(safeTitle)
	if libraryRoot == "" {
		return safeTitle + "/" + seasonDir
	}
	sep := "/"
	if strings.Contains(libraryRoot, `\`) && !strings.Contains(libraryRoot, "/") {
		sep = `\`
	}
	return strings.TrimRight(libraryRoot, `/\`) + sep + safeTitle + sep + seasonDir
}

func PlanEpisodeRenames(displayName string, files []qb.File) []Plan {
	usable := make([]qb.File, 0, len(files))
	for _, file := range files {
		if isSkippedExtra(file.Name) {
			continue
		}
		usable = append(usable, file)
	}
	var videos []qb.File
	for _, file := range usable {
		if _, ok := videoExt[extname(file.Name)]; ok {
			videos = append(videos, file)
		}
	}
	if len(videos) == 0 {
		return nil
	}
	primary := videos[0]
	for _, file := range videos[1:] {
		if file.Size > primary.Size {
			primary = file
		}
	}
	primaryStem := stem(primary.Name)
	dir := dirname(primary.Name)
	var plans []Plan
	videoTo := joinDir(dir, displayName+extname(primary.Name))
	if primary.Name != videoTo {
		plans = append(plans, Plan{From: primary.Name, To: videoTo})
	}
	for _, file := range usable {
		if _, ok := subExt[extname(file.Name)]; !ok {
			continue
		}
		subStem := stem(file.Name)
		if subStem != primaryStem && !strings.HasPrefix(subStem, primaryStem) {
			continue
		}
		extra := strings.TrimPrefix(subStem, primaryStem)
		subTo := joinDir(dir, displayName+extra+extname(file.Name))
		if file.Name != subTo {
			plans = append(plans, Plan{From: file.Name, To: subTo})
		}
	}
	return plans
}

func isSkippedExtra(filePath string) bool {
	for _, part := range strings.FieldsFunc(filePath, func(r rune) bool {
		return r == '/' || r == '\\'
	}) {
		if extraToken.MatchString(part) {
			return true
		}
	}
	return false
}

func extname(filePath string) string {
	base := basename(filePath)
	i := strings.LastIndex(base, ".")
	if i < 0 {
		return ""
	}
	return strings.ToLower(base[i:])
}

func basename(filePath string) string {
	return path.Base(strings.ReplaceAll(filePath, `\`, "/"))
}

func dirname(filePath string) string {
	normalized := strings.ReplaceAll(filePath, `\`, "/")
	dir := path.Dir(normalized)
	if dir == "." {
		return ""
	}
	if strings.Contains(filePath, `\`) && !strings.Contains(filePath, "/") {
		return strings.ReplaceAll(dir, "/", `\`)
	}
	return dir
}

func stem(filePath string) string {
	base := basename(filePath)
	i := strings.LastIndex(base, ".")
	if i < 0 {
		return base
	}
	return base[:i]
}

func joinDir(dir, name string) string {
	if dir == "" {
		return name
	}
	sep := "/"
	if strings.Contains(dir, `\`) && !strings.Contains(dir, "/") {
		sep = `\`
	}
	return dir + sep + name
}
